adoc/extensions/sycl_khr_queue_empty_query.adoc

```diff
@@ -0,0 +1,81 @@
+[[sec:khr-queue-empty-query]]
+= sycl_khr_queue_empty_query
+
+This extension allows developers to query the queue's emptiness, meaning if all
+commands submitted to a queue have been completed.
+
+[[sec:khr-queue-empty-query-dependencies]]
+== Dependencies
+
+This extension has no dependencies on other extensions.
+
+[[sec:khr-queue-empty-query-feature-test]]
+== Feature test macro
+
+An implementation supporting this extension must predefine the macro
+[code]#SYCL_KHR_QUEUE_EMPTY_QUERY# to one of the values defined in the table
+below.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+[[sec:khr-queue-empty-query-funct]]
+== New Queue Function to Query Emptiness
+
+This extension adds the following function to the [code]#sycl::queue# class,
+which provides information about the emptiness of the queue.
+
+'''
+
+.[apidef]#queue::khr_empty#
+[source,role=synopsis,id=api:queue-khr-empty]
+----
+bool khr_empty() const
+----
+
+_Synchronization_: When this function returns [code]#true#, equivalent to
+[code]#queue::wait()#.
+
+_Returns:_ [code]#true# if all <<command,commands>> enqueued on this queue have
```

Comment:
Should we perhaps add a note to say that repeatedly calling `khr_empty()` does not guarantee progress on its own? From the discussion below I think there is agreement that SYCL implementations might execute commands lazily such that such patterns may not work.

I'm a bit concerned that the equivalence to `queue::wait()` might be misinterpreted by users as providing similar progress guarantees.
I know that we're not saying this here, it's just that I wouldn't be surprised if users make this assumption and feel confirmed when they read what we write.

Reply:
Good idea. 

Yep, this lazy / greepy discussion belong to a new KHR (adding querry / property of the queue or something)

Will work on phrasing! 

Thanks!

Reply:
We could say something like:
> _Synchronization_: When this function returns `true`, synchronization behavior is equivalent to `queue::wait()`.

...but this seems unnecessarily verbose to me, because the whole paragraph is about synchronization behavior.

If we want to add a note about the lack of progress, I think we should add a more general note to the core specification.  Repeatedly calling `khr_empty()` is the same as repeatedly checking the status of one or more `sycl::event` objects, but I don't think we explicitly say anywhere that querying an event's status does not guarantee progress.

Reply:
> If we want to add a note about the lack of progress, I think we should add a more general note to the core specification. Repeatedly calling khr_empty() is the same as repeatedly checking the status of one or more sycl::event objects, but I don't think we explicitly say anywhere that querying an event's status does not guarantee progress.

I agree that it's the same as repeatedly checking the completion status of an event, and I'm also not aware that we say something there, *but* we also don't draw any comparisons to `queue::wait()` there either. This is the important bit in my opinion.

In any case, this is not a hill I'm willing to die on, just a suggestion to avoid misinterpretation of the spec by users.

Reply:
I agree with you both. We should put a note somewhere on the "greedy / lazy -ness"  and that "querry status"  doesn't trigger a submission. 

(like we did for free, and blocking). 

I will need to think more about where it makes sense to put it!  

The `while Q.empy` is the "traditional" CUDA pattern, so it makes sense to put it here. But the problem is more general, so make sense to also put it on the spec :')

Reply:
@illuhad I put a simple note to say that it works best with `khr_flush`. So people can read it to get more information.
Not perfect, but it was too hard to write a sentence to explain the greedy/lazy in this KHR for my little brain.

It was a good point, that this this is mostly useless / implementation speficit,  without flush